sa/model.go

```diff
@@ -1294,3 +1294,245 @@ func setReplacementOrderFinalized(ctx context.Context, db db.Execer, orderID int
 	}
 	return nil
 }
+
+type pauseId struct {
+	IdentifierType  uint8  `db:"identifierType"`
+	IdentifierValue string `db:"identifierValue"`
+}
+
+func newIdentifierFromPB(pb *sapb.Identifier) (pauseId, error) {
+	idType, ok := identifierTypeToUint[pb.Type]
+	if !ok {
+		return pauseId{}, fmt.Errorf("unsupported identifier type %q", pb.Type)
+	}
+
+	return pauseId{
+		IdentifierType:  idType,
+		IdentifierValue: pb.Value,
+	}, nil
+}
+
+type pauseIds []pauseId
+
+func newIdentifiersFromPB(pb []*sapb.Identifier) (pauseIds, error) {
+	var ids pauseIds
+	for _, p := range pb {
+		id, err := newIdentifierFromPB(p)
+		if err != nil {
+			return nil, err
+		}
+		ids = append(ids, id)
+	}
+	return ids, nil
+}
+
+// identifierValuesForType returns a list of identifier values for the given
+// identifier type. If no values are found, an empty list is returned.
+func (id pauseIds) identifierValuesForType(identifierType uint8) []string {
+	var values []string
+	for _, i := range id {
+		if i.IdentifierType == identifierType {
+			values = append(values, i.IdentifierValue)
+		}
+	}
+	return values
+}
+
+// pausedModel represents a row in the paused table. The pausedAt and unpausedAt
+// fields are pointers because they are NULL-able columns. At no point should
+// both pausedAt and unpausedAt be non-NULL. Valid states are:
+//   - identifier paused: pausedAt is non-NULL, unpausedAt is NULL
+//   - identifier unpaused: pausedAt is non-NULL, unpausedAt is non-NULL
+type pausedModel struct {
+	pauseId
+	RegistrationID int64      `db:"registrationID"`
+	PausedAt       *time.Time `db:"pausedAt"`
+	UnpausedAt     *time.Time `db:"unpausedAt"`
+}
+
+// checkIdentifiersPaused takes a slice of identifiers of type "dns" and returns
+// a slice of the first 15 identifier values which are currently paused for the
+// provided account. If no matches are found, an empty slice is returned. If a
+// non-DNS identifier is provided, an error is returned.
+func checkIdentifiersPaused(ctx context.Context, dbs db.Selector, regID int64, ids pauseIds) ([]string, error) {
```

Comment:
Why is this function (and the similar complex helper functions below) in this file? These are each called in only one place, and they have no direct tests (all of the tests operate at the SA level). Why not inline their logic directly into the SA methods? This would also prevent having to deal with the annoyance of ensuring the executor they're called with is actually a transaction.

Reply:
This is fair. My original approach was to test these directly but once I fleshed out the SA methods that changed.